pkg/api/api.go

```diff
@@ -354,12 +367,73 @@ func (a *API) RegisterStoreGateway(s *storegateway.StoreGateway) {
 	a.RegisterRoute("/store-gateway/tenant/{tenant}/blocks", http.HandlerFunc(s.BlocksHandler), false, true, "GET")
 }
 
-// RegisterCompactor registers the ring UI page associated with the compactor.
+// RegisterCompactor registers routes associated with the compactor.
 func (a *API) RegisterCompactor(c *compactor.MultitenantCompactor) {
 	a.indexPage.AddLinks(defaultWeight, "Compactor", []IndexPageLink{
 		{Desc: "Ring status", Path: "/compactor/ring"},
 	})
 	a.RegisterRoute("/compactor/ring", http.HandlerFunc(c.RingHandler), false, true, "GET", "POST")
+	// Endpoint to handle requests for completion of block backfilling.
+	// Placed before less specific handler for same route, that doesn't take query parameter, for precedence.
+	a.RegisterRouteWithQueryParameters("/api/v1/upload/block/{block}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		blockID := vars["block"]
+
+		tenantID, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+		if err != nil {
+			http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+			return
+		}
+
+		if err := c.CompleteBlockUpload(ctx, tenantID, blockID, r); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+	}), true, false, []string{"uploadComplete", "true"}, http.MethodPost)
+	// Endpoint to handle requests for starting of block backfilling.
+	// Starting the backfilling of a block means to verify that the backfill can go ahead.
+	// In practice this means to check that the block isn't already in block storage.
+	a.RegisterRoute("/api/v1/upload/block/{block}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		blockID := vars["block"]
+		tenantID, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+		if err != nil {
+			http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+			return
+		}
+
+		if err := c.CreateBlockUpload(ctx, tenantID, blockID); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+	}), true, false, http.MethodPost)
+	// Endpoint to handle requests for uploading block files to a backfill.
+	a.RegisterRoute("/api/v1/upload/block/{block}/{path}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		blockID := vars["block"]
+		pth, err := url.PathUnescape(vars["path"])
+		if err != nil {
+			http.Error(w, fmt.Sprintf("invalid file path: %q", vars["path"]), http.StatusBadRequest)
+			return
+		}
+
+		tenantID, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+		if err != nil {
+			http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+			return
+		}
+
+		if err := c.UploadBlockFile(ctx, tenantID, blockID, pth, r); err != nil {
+			http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+	}), true, false, http.MethodPost)
```

Comment:
IMO this handler code should go in the MultitenantCompactor

Reply:
@colega yep, I'll clean up before moving the PR out of draft mode.

Reply:
Done.